File: Cpp/algo_codes/solu_cpp/follow_carl/DP/integerBreak.cc
#include<iostream>
#include<vector>

using std::cout;
using std::endl;
using std::vector;

int max(int a,int b){
	return a >= b?a:b;
}

int max(int a , int b ,int c){
	return max(a,b) >= c ? max(a,b) : c;
}

int integerBreak(int n){
	vector<int> dp(n+1);
	dp[0] = 0;
	dp[1] = 0;
	dp[2] = 1;
	for(int i = 3;i<n+1;i++){
		for(int j = 1 ; j<=i/2;j++){
			dp[i] = max(j*dp[i-j],j*(i-j),dp[i]);
			cout<<i<<":" <<dp[i]<<endl;
		}
	}
	cout<<dp[3]<<endl;
	return dp[n];

}

int main(){
	int n = 10;
	cout<<integerBreak(n)<<endl;
	return 0;
}

File: Python/algo_codes/solu_py/follow_carl/DP/integerBreak.py
# def integerBreak(self,n:int)->int:
#     dp = [0]*(n+1)
#     dp[2] = 1
#     for i in range(3,n+1):
#         for j in range(1,i//2 +1):
#             dp[i] = max(j*(i-j),j*dp[i-j],dp[i])

#     return dp[n]

def integerBreak(self, n: int) -> int:
    if n <= 3:
        return n-1

    q, r = n//3, n % 3
    if r == 0:
        return 3**q
    elif r == 1:
        return 3**(q-1)*4
    else:
        return 3**q*2


n = 10
print(integerBreak(None, n))
